core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1350,7 +1352,11 @@ class Log(@volatile var dir: File,
       }
 
       // check the validity of the message by checking CRC
-      batch.ensureValid()
+      if (!batch.isValid) {
+        brokerTopicStats.topicStats(topicPartition.topic).invalidMessageCrcRecordsPerSec.mark()
+        brokerTopicStats.allTopicsStats.invalidMessageCrcRecordsPerSec.mark()
+        throw new InvalidRecordException(s"Record is corrupt (stored crc = ${batch.checksum()}).")
```

Comment:
Also add topic / partition into the error message of this invalidMessageCrc since we only record at all-topics level now, so we rely on the WARN log4j entry to know such finer grained information.